File: problems/图/695.岛屿的最大面积.py
#!/usr/bin/env python3
"""
695. 岛屿的最大面积
"""
def max_area_of_island(grid: list[list[int]]) -> int:
    m, n = len(grid), len(grid[0])

    def dfs(r, c):
        if r < 0 or r >= m or c < 0 or c >= n or grid[r][c] == 0:
            return 0
        grid[r][c] = 0
        return 1 + dfs(r+1, c) + dfs(r-1, c) + dfs(r, c+1) + dfs(r, c-1)

    max_area = 0
    for r in range(m):
        for c in range(n):
            if grid[r][c] == 1:
                max_area = max(max_area, dfs(r, c))
    return max_area

if __name__ == "__main__":
    # 测试用例
    grid1 = [[0, 0, 1, 0, 0, 0, 0, 1, 0, 0], [0, 0, 0, 0, 0, 0, 0, 1, 1, 1], [0, 1, 1, 0, 1, 0, 0, 0, 0, 0], [0, 1, 0, 0, 1, 1, 0, 0, 1, 0]]
    print(max_area_of_island(grid1))  # 预期: 6
    grid2 = [[0, 0, 0, 0, 0, 0, 0, 0]]
    print(max_area_of_island(grid2))  # 预期: 0
    grid3 = [[1, 1, 0, 0, 0], [1, 1, 0, 0, 0], [0, 0, 0, 1, 1], [0, 0, 0, 1, 1]]
    print(max_area_of_island(grid3))  # 预期: 4

File: problems/图/695.岛屿的最大面积.cpp
#include <iostream>
#include <vector>
#include <utility>
using namespace std;

int maxAreaOfIsland(vector<vector<int>>& grid) {
    int m = grid.size(), n = grid[0].size(), area = 0;
    function<int(int,int)> dfs = [&](int r, int c) -> int {
        if (r < 0 || r >= m || c < 0 || c >= n || grid[r][c] == 0) return 0;
        grid[r][c] = 0;
        return 1 + dfs(r+1, c) + dfs(r-1, c) + dfs(r, c+1) + dfs(r, c-1);
    };
    for (int r = 0; r < m; r++)
        for (int c = 0; c < n; c++)
            if (grid[r][c]) area = max(area, dfs(r, c));
    return area;
}

int main() {
    // 测试用例
    vector<vector<int>> grid1 = {{0,0,1,0,0,0,0,1,0,0},{0,0,0,0,0,0,0,1,1,1},{0,1,1,0,1,0,0,0,0,0},{0,1,0,0,1,1,0,0,1,0}};
    cout << maxAreaOfIsland(grid1) << endl;  // 预期: 6
    vector<vector<int>> grid2 = {{0,0,0,0,0,0,0,0}};
    cout << maxAreaOfIsland(grid2) << endl;  // 预期: 0
    vector<vector<int>> grid3 = {{1,1,0,0,0},{1,1,0,0,0},{0,0,0,1,1},{0,0,0,1,1}};
    cout << maxAreaOfIsland(grid3) << endl;  // 预期: 4
    return 0;
}
